// @cornerstonejs/tools/dist/cjs/utilities/planar/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterAnnotationsWithinSamePlane = exports.isPlaneIntersectingAABB = exports.getPointInLineOfSightWithCriteria = exports.filterAnnotationsForDisplay = exports.getWorldWidthAndHeightFromCorners = exports.filterAnnotationsWithinSlice = void 0;
const filterAnnotationsWithinSlice_1 = __importDefault(require("./filterAnnotationsWithinSlice"));
exports.filterAnnotationsWithinSlice = filterAnnotationsWithinSlice_1.default;
const getWorldWidthAndHeightFromCorners_1 = __importDefault(require("./getWorldWidthAndHeightFromCorners"));
exports.getWorldWidthAndHeightFromCorners = getWorldWidthAndHeightFromCorners_1.default;
const filterAnnotationsForDisplay_1 = __importDefault(require("./filterAnnotationsForDisplay"));
exports.filterAnnotationsForDisplay = filterAnnotationsForDisplay_1.default;
const getPointInLineOfSightWithCriteria_1 = __importDefault(require("./getPointInLineOfSightWithCriteria"));
exports.getPointInLineOfSightWithCriteria = getPointInLineOfSightWithCriteria_1.default;
const isPlaneIntersectingAABB_1 = require("./isPlaneIntersectingAABB");
Object.defineProperty(exports, "isPlaneIntersectingAABB", { enumerable: true, get: function () { return isPlaneIntersectingAABB_1.isPlaneIntersectingAABB; } });
const filterAnnotationsWithinPlane_1 = require("./filterAnnotationsWithinPlane");
Object.defineProperty(exports, "filterAnnotationsWithinSamePlane", { enumerable: true, get: function () { return filterAnnotationsWithinPlane_1.filterAnnotationsWithinSamePlane; } });
exports.default = {
    filterAnnotationsWithinSlice: filterAnnotationsWithinSlice_1.default,
    getWorldWidthAndHeightFromCorners: getWorldWidthAndHeightFromCorners_1.default,
    filterAnnotationsForDisplay: filterAnnotationsForDisplay_1.default,
    getPointInLineOfSightWithCriteria: getPointInLineOfSightWithCriteria_1.default,
    isPlaneIntersectingAABB: isPlaneIntersectingAABB_1.isPlaneIntersectingAABB,
    filterAnnotationsWithinSamePlane: filterAnnotationsWithinPlane_1.filterAnnotationsWithinSamePlane,
};
