// @cornerstonejs/core/dist/cjs/utilities/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCurrentVolumeViewportSlice = exports.getClosestStackImageIndexForPoint = exports.getViewportsWithImageURI = exports.actorIsA = exports.isPTPrescaledWithSUV = exports.isImageActor = exports.getImageSliceDataForVolumeViewport = exports.snapFocalPointToSlice = exports.getSliceRange = exports.getVolumeViewportScrollInfo = exports.getVolumeSliceRangeInfo = exports.imageToWorldCoords = exports.worldToImageCoords = exports.renderToCanvasGPU = exports.renderToCanvasCPU = exports.loadImageToCanvas = exports.transformIndexToWorld = exports.transformWorldToIndex = exports.getViewportsWithVolumeId = exports.getVolumeViewportsContainingSameVolumes = exports.indexWithinDimensions = exports.getVolumeActorCorners = exports.getTargetVolumeAndSpacingInNormalDir = exports.getSpacingInNormalDirection = exports.getClosestImageId = exports.convertToGrayscale = exports.windowLevel = exports.getViewportModality = exports.createInt16SharedArray = exports.createUint16SharedArray = exports.createUint8SharedArray = exports.createFloat32SharedArray = exports.isOpposite = exports.isEqualNegative = exports.isEqualAbs = exports.isEqual = exports.getRuntimeId = exports.getMinMax = exports.planar = exports.uuidv4 = exports.clamp = exports.calibratedPixelSpacingMetadataProvider = exports.imageIdToURI = exports.triggerEvent = exports.scaleRgbTransferFunction = exports.createLinearRGBTransferFunction = exports.getVoiFromSigmoidRGBTransferFunction = exports.createSigmoidRGBTransferFunction = exports.invertRgbTransferFunction = exports.eventListener = void 0;
exports.hasFloatScalingParameters = exports.color = exports.getVolumeId = exports.getRandomSampleFromArray = exports.getViewportImageIds = exports.roundToPrecision = exports.roundNumber = exports.cacheUtils = exports.convertVolumeToStackViewport = exports.convertStackToVolumeViewport = exports.RLEVoxelMap = exports.VoxelManager = exports.getBufferConfiguration = exports.generateVolumePropsFromImageIds = exports.isVideoTransferSyntax = exports.genericMetadataProvider = exports.isValidVolume = exports.makeVolumeMetadata = exports.sortImageIdsAndGetSpacing = exports.updateVTKImageDataWithCornerstoneImage = exports.transferFunctionUtils = exports.imageRetrieveMetadataProvider = exports.decimate = exports.ProgressiveIterator = exports.getImageLegacy = exports.colormap = exports.getScalarDataType = exports.getScalingParameters = exports.PointsManager = exports.deepMerge = exports.applyPreset = exports.hasNaNValues = exports.getViewportImageCornersInWorld = exports.spatialRegistrationMetadataProvider = exports.calculateViewportsSpatialRegistration = void 0;
const eventListener = __importStar(require("./eventListener"));
exports.eventListener = eventListener;
const invertRgbTransferFunction_1 = __importDefault(require("./invertRgbTransferFunction"));
exports.invertRgbTransferFunction = invertRgbTransferFunction_1.default;
const createSigmoidRGBTransferFunction_1 = __importDefault(require("./createSigmoidRGBTransferFunction"));
exports.createSigmoidRGBTransferFunction = createSigmoidRGBTransferFunction_1.default;
const getVoiFromSigmoidRGBTransferFunction_1 = __importDefault(require("./getVoiFromSigmoidRGBTransferFunction"));
exports.getVoiFromSigmoidRGBTransferFunction = getVoiFromSigmoidRGBTransferFunction_1.default;
const createLinearRGBTransferFunction_1 = __importDefault(require("./createLinearRGBTransferFunction"));
exports.createLinearRGBTransferFunction = createLinearRGBTransferFunction_1.default;
const scaleRgbTransferFunction_1 = __importDefault(require("./scaleRgbTransferFunction"));
exports.scaleRgbTransferFunction = scaleRgbTransferFunction_1.default;
const triggerEvent_1 = __importDefault(require("./triggerEvent"));
exports.triggerEvent = triggerEvent_1.default;
const uuidv4_1 = __importDefault(require("./uuidv4"));
exports.uuidv4 = uuidv4_1.default;
const getMinMax_1 = __importDefault(require("./getMinMax"));
exports.getMinMax = getMinMax_1.default;
const getRuntimeId_1 = __importDefault(require("./getRuntimeId"));
exports.getRuntimeId = getRuntimeId_1.default;
const imageIdToURI_1 = __importDefault(require("./imageIdToURI"));
exports.imageIdToURI = imageIdToURI_1.default;
const calibratedPixelSpacingMetadataProvider_1 = __importDefault(require("./calibratedPixelSpacingMetadataProvider"));
exports.calibratedPixelSpacingMetadataProvider = calibratedPixelSpacingMetadataProvider_1.default;
const clamp_1 = __importDefault(require("./clamp"));
exports.clamp = clamp_1.default;
const isEqual_1 = require("./isEqual");
Object.defineProperty(exports, "isEqual", { enumerable: true, get: function () { return isEqual_1.isEqual; } });
Object.defineProperty(exports, "isEqualAbs", { enumerable: true, get: function () { return isEqual_1.isEqualAbs; } });
Object.defineProperty(exports, "isEqualNegative", { enumerable: true, get: function () { return isEqual_1.isEqualNegative; } });
const isOpposite_1 = __importDefault(require("./isOpposite"));
exports.isOpposite = isOpposite_1.default;
const createUint8SharedArray_1 = __importDefault(require("./createUint8SharedArray"));
exports.createUint8SharedArray = createUint8SharedArray_1.default;
const createFloat32SharedArray_1 = __importDefault(require("./createFloat32SharedArray"));
exports.createFloat32SharedArray = createFloat32SharedArray_1.default;
const createUInt16SharedArray_1 = __importDefault(require("./createUInt16SharedArray"));
exports.createUint16SharedArray = createUInt16SharedArray_1.default;
const createInt16SharedArray_1 = __importDefault(require("./createInt16SharedArray"));
exports.createInt16SharedArray = createInt16SharedArray_1.default;
const getViewportModality_1 = __importDefault(require("./getViewportModality"));
exports.getViewportModality = getViewportModality_1.default;
const getClosestImageId_1 = __importDefault(require("./getClosestImageId"));
exports.getClosestImageId = getClosestImageId_1.default;
const getSpacingInNormalDirection_1 = __importDefault(require("./getSpacingInNormalDirection"));
exports.getSpacingInNormalDirection = getSpacingInNormalDirection_1.default;
const getTargetVolumeAndSpacingInNormalDir_1 = __importDefault(require("./getTargetVolumeAndSpacingInNormalDir"));
exports.getTargetVolumeAndSpacingInNormalDir = getTargetVolumeAndSpacingInNormalDir_1.default;
const getVolumeActorCorners_1 = __importDefault(require("./getVolumeActorCorners"));
exports.getVolumeActorCorners = getVolumeActorCorners_1.default;
const indexWithinDimensions_1 = __importDefault(require("./indexWithinDimensions"));
exports.indexWithinDimensions = indexWithinDimensions_1.default;
const getVolumeViewportsContainingSameVolumes_1 = __importDefault(require("./getVolumeViewportsContainingSameVolumes"));
exports.getVolumeViewportsContainingSameVolumes = getVolumeViewportsContainingSameVolumes_1.default;
const getViewportsWithVolumeId_1 = __importDefault(require("./getViewportsWithVolumeId"));
exports.getViewportsWithVolumeId = getViewportsWithVolumeId_1.default;
const transformWorldToIndex_1 = __importDefault(require("./transformWorldToIndex"));
exports.transformWorldToIndex = transformWorldToIndex_1.default;
const transformIndexToWorld_1 = __importDefault(require("./transformIndexToWorld"));
exports.transformIndexToWorld = transformIndexToWorld_1.default;
const loadImageToCanvas_1 = __importDefault(require("./loadImageToCanvas"));
exports.loadImageToCanvas = loadImageToCanvas_1.default;
const renderToCanvasCPU_1 = __importDefault(require("./renderToCanvasCPU"));
exports.renderToCanvasCPU = renderToCanvasCPU_1.default;
const renderToCanvasGPU_1 = __importDefault(require("./renderToCanvasGPU"));
exports.renderToCanvasGPU = renderToCanvasGPU_1.default;
const worldToImageCoords_1 = __importDefault(require("./worldToImageCoords"));
exports.worldToImageCoords = worldToImageCoords_1.default;
const imageToWorldCoords_1 = __importDefault(require("./imageToWorldCoords"));
exports.imageToWorldCoords = imageToWorldCoords_1.default;
const getVolumeSliceRangeInfo_1 = __importDefault(require("./getVolumeSliceRangeInfo"));
exports.getVolumeSliceRangeInfo = getVolumeSliceRangeInfo_1.default;
const getVolumeViewportScrollInfo_1 = __importDefault(require("./getVolumeViewportScrollInfo"));
exports.getVolumeViewportScrollInfo = getVolumeViewportScrollInfo_1.default;
const getSliceRange_1 = __importDefault(require("./getSliceRange"));
exports.getSliceRange = getSliceRange_1.default;
const snapFocalPointToSlice_1 = __importDefault(require("./snapFocalPointToSlice"));
exports.snapFocalPointToSlice = snapFocalPointToSlice_1.default;
const getImageSliceDataForVolumeViewport_1 = __importDefault(require("./getImageSliceDataForVolumeViewport"));
exports.getImageSliceDataForVolumeViewport = getImageSliceDataForVolumeViewport_1.default;
const actorCheck_1 = require("./actorCheck");
Object.defineProperty(exports, "isImageActor", { enumerable: true, get: function () { return actorCheck_1.isImageActor; } });
Object.defineProperty(exports, "actorIsA", { enumerable: true, get: function () { return actorCheck_1.actorIsA; } });
const getViewportsWithImageURI_1 = __importDefault(require("./getViewportsWithImageURI"));
exports.getViewportsWithImageURI = getViewportsWithImageURI_1.default;
const getClosestStackImageIndexForPoint_1 = __importDefault(require("./getClosestStackImageIndexForPoint"));
exports.getClosestStackImageIndexForPoint = getClosestStackImageIndexForPoint_1.default;
const getCurrentVolumeViewportSlice_1 = __importDefault(require("./getCurrentVolumeViewportSlice"));
exports.getCurrentVolumeViewportSlice = getCurrentVolumeViewportSlice_1.default;
const calculateViewportsSpatialRegistration_1 = __importDefault(require("./calculateViewportsSpatialRegistration"));
exports.calculateViewportsSpatialRegistration = calculateViewportsSpatialRegistration_1.default;
const spatialRegistrationMetadataProvider_1 = __importDefault(require("./spatialRegistrationMetadataProvider"));
exports.spatialRegistrationMetadataProvider = spatialRegistrationMetadataProvider_1.default;
const getViewportImageCornersInWorld_1 = __importDefault(require("./getViewportImageCornersInWorld"));
exports.getViewportImageCornersInWorld = getViewportImageCornersInWorld_1.default;
const hasNaNValues_1 = __importDefault(require("./hasNaNValues"));
exports.hasNaNValues = hasNaNValues_1.default;
const applyPreset_1 = __importDefault(require("./applyPreset"));
exports.applyPreset = applyPreset_1.default;
const PointsManager_1 = __importDefault(require("./PointsManager"));
exports.PointsManager = PointsManager_1.default;
const deepMerge_1 = __importDefault(require("./deepMerge"));
exports.deepMerge = deepMerge_1.default;
const getScalingParameters_1 = __importDefault(require("./getScalingParameters"));
exports.getScalingParameters = getScalingParameters_1.default;
const getScalarDataType_1 = __importDefault(require("./getScalarDataType"));
exports.getScalarDataType = getScalarDataType_1.default;
const isPTPrescaledWithSUV_1 = __importDefault(require("./isPTPrescaledWithSUV"));
exports.isPTPrescaledWithSUV = isPTPrescaledWithSUV_1.default;
const getImageLegacy_1 = __importDefault(require("./getImageLegacy"));
exports.getImageLegacy = getImageLegacy_1.default;
const sortImageIdsAndGetSpacing_1 = __importDefault(require("./sortImageIdsAndGetSpacing"));
exports.sortImageIdsAndGetSpacing = sortImageIdsAndGetSpacing_1.default;
const makeVolumeMetadata_1 = __importDefault(require("./makeVolumeMetadata"));
exports.makeVolumeMetadata = makeVolumeMetadata_1.default;
const genericMetadataProvider_1 = __importDefault(require("./genericMetadataProvider"));
exports.genericMetadataProvider = genericMetadataProvider_1.default;
const isValidVolume_1 = require("./isValidVolume");
Object.defineProperty(exports, "isValidVolume", { enumerable: true, get: function () { return isValidVolume_1.isValidVolume; } });
const updateVTKImageDataWithCornerstoneImage_1 = require("./updateVTKImageDataWithCornerstoneImage");
Object.defineProperty(exports, "updateVTKImageDataWithCornerstoneImage", { enumerable: true, get: function () { return updateVTKImageDataWithCornerstoneImage_1.updateVTKImageDataWithCornerstoneImage; } });
const ProgressiveIterator_1 = __importDefault(require("./ProgressiveIterator"));
exports.ProgressiveIterator = ProgressiveIterator_1.default;
const decimate_1 = __importDefault(require("./decimate"));
exports.decimate = decimate_1.default;
const imageRetrieveMetadataProvider_1 = __importDefault(require("./imageRetrieveMetadataProvider"));
exports.imageRetrieveMetadataProvider = imageRetrieveMetadataProvider_1.default;
const isVideoTransferSyntax_1 = __importDefault(require("./isVideoTransferSyntax"));
exports.isVideoTransferSyntax = isVideoTransferSyntax_1.default;
const getBufferConfiguration_1 = require("./getBufferConfiguration");
Object.defineProperty(exports, "getBufferConfiguration", { enumerable: true, get: function () { return getBufferConfiguration_1.getBufferConfiguration; } });
const generateVolumePropsFromImageIds_1 = require("./generateVolumePropsFromImageIds");
Object.defineProperty(exports, "generateVolumePropsFromImageIds", { enumerable: true, get: function () { return generateVolumePropsFromImageIds_1.generateVolumePropsFromImageIds; } });
const convertStackToVolumeViewport_1 = require("./convertStackToVolumeViewport");
Object.defineProperty(exports, "convertStackToVolumeViewport", { enumerable: true, get: function () { return convertStackToVolumeViewport_1.convertStackToVolumeViewport; } });
const convertVolumeToStackViewport_1 = require("./convertVolumeToStackViewport");
Object.defineProperty(exports, "convertVolumeToStackViewport", { enumerable: true, get: function () { return convertVolumeToStackViewport_1.convertVolumeToStackViewport; } });
const VoxelManager_1 = __importDefault(require("./VoxelManager"));
exports.VoxelManager = VoxelManager_1.default;
const RLEVoxelMap_1 = __importDefault(require("./RLEVoxelMap"));
exports.RLEVoxelMap = RLEVoxelMap_1.default;
const roundNumber_1 = __importStar(require("./roundNumber"));
exports.roundNumber = roundNumber_1.default;
Object.defineProperty(exports, "roundToPrecision", { enumerable: true, get: function () { return roundNumber_1.roundToPrecision; } });
const convertToGrayscale_1 = __importDefault(require("./convertToGrayscale"));
exports.convertToGrayscale = convertToGrayscale_1.default;
const getViewportImageIds_1 = __importDefault(require("./getViewportImageIds"));
exports.getViewportImageIds = getViewportImageIds_1.default;
const getRandomSampleFromArray_1 = require("./getRandomSampleFromArray");
Object.defineProperty(exports, "getRandomSampleFromArray", { enumerable: true, get: function () { return getRandomSampleFromArray_1.getRandomSampleFromArray; } });
const getVolumeId_1 = require("./getVolumeId");
Object.defineProperty(exports, "getVolumeId", { enumerable: true, get: function () { return getVolumeId_1.getVolumeId; } });
const hasFloatScalingParameters_1 = require("./hasFloatScalingParameters");
Object.defineProperty(exports, "hasFloatScalingParameters", { enumerable: true, get: function () { return hasFloatScalingParameters_1.hasFloatScalingParameters; } });
const planar = __importStar(require("./planar"));
exports.planar = planar;
const windowLevel = __importStar(require("./windowLevel"));
exports.windowLevel = windowLevel;
const colormap = __importStar(require("./colormap"));
exports.colormap = colormap;
const transferFunctionUtils = __importStar(require("./transferFunctionUtils"));
exports.transferFunctionUtils = transferFunctionUtils;
const cacheUtils = __importStar(require("./cacheUtils"));
exports.cacheUtils = cacheUtils;
const color = __importStar(require("./color"));
exports.color = color;
