// @cornerstonejs/tools/dist/cjs/utilities/contours/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculatePerimeter = exports.findHandlePolylineIndex = exports.acceptAutogeneratedInterpolations = exports.interpolation = exports.updateContourPolyline = exports.getContourHolesDataCanvas = exports.getContourHolesDataWorld = exports.AnnotationToPointData = exports.generateContourSetsFromLabelmap = exports.detectContourHoles = exports.getDeduplicatedVTKPolyDataPoints = exports.contourFinder = exports.areCoplanarContours = void 0;
const areCoplanarContours_1 = __importDefault(require("./areCoplanarContours"));
exports.areCoplanarContours = areCoplanarContours_1.default;
const contourFinder_1 = __importDefault(require("./contourFinder"));
exports.contourFinder = contourFinder_1.default;
const getDeduplicatedVTKPolyDataPoints_1 = require("./getDeduplicatedVTKPolyDataPoints");
Object.defineProperty(exports, "getDeduplicatedVTKPolyDataPoints", { enumerable: true, get: function () { return getDeduplicatedVTKPolyDataPoints_1.getDeduplicatedVTKPolyDataPoints; } });
const detectContourHoles_1 = __importDefault(require("./detectContourHoles"));
exports.detectContourHoles = detectContourHoles_1.default;
const generateContourSetsFromLabelmap_1 = require("./generateContourSetsFromLabelmap");
Object.defineProperty(exports, "generateContourSetsFromLabelmap", { enumerable: true, get: function () { return generateContourSetsFromLabelmap_1.generateContourSetsFromLabelmap; } });
const AnnotationToPointData_1 = __importDefault(require("./AnnotationToPointData"));
exports.AnnotationToPointData = AnnotationToPointData_1.default;
const getContourHolesDataWorld_1 = __importDefault(require("./getContourHolesDataWorld"));
exports.getContourHolesDataWorld = getContourHolesDataWorld_1.default;
const getContourHolesDataCanvas_1 = __importDefault(require("./getContourHolesDataCanvas"));
exports.getContourHolesDataCanvas = getContourHolesDataCanvas_1.default;
const updateContourPolyline_1 = __importDefault(require("./updateContourPolyline"));
exports.updateContourPolyline = updateContourPolyline_1.default;
const acceptAutogeneratedInterpolations_1 = __importDefault(require("./interpolation/acceptAutogeneratedInterpolations"));
exports.acceptAutogeneratedInterpolations = acceptAutogeneratedInterpolations_1.default;
const interpolation = __importStar(require("./interpolation"));
exports.interpolation = interpolation;
const findHandlePolylineIndex_1 = __importDefault(require("./findHandlePolylineIndex"));
exports.findHandlePolylineIndex = findHandlePolylineIndex_1.default;
const calculatePerimeter_1 = __importDefault(require("./calculatePerimeter"));
exports.calculatePerimeter = calculatePerimeter_1.default;
