// @cornerstonejs/tools/dist/cjs/store/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Synchronizer = exports.SynchronizerManager = exports.ToolGroupManager = exports.svgNodeCache = exports.cancelActiveManipulations = exports.removeEnabledElement = exports.addEnabledElement = exports.removeTool = exports.addTool = exports.state = void 0;
const addTool_1 = require("./addTool");
Object.defineProperty(exports, "addTool", { enumerable: true, get: function () { return addTool_1.addTool; } });
Object.defineProperty(exports, "removeTool", { enumerable: true, get: function () { return addTool_1.removeTool; } });
const addEnabledElement_1 = __importDefault(require("./addEnabledElement"));
exports.addEnabledElement = addEnabledElement_1.default;
const removeEnabledElement_1 = __importDefault(require("./removeEnabledElement"));
exports.removeEnabledElement = removeEnabledElement_1.default;
const cancelActiveManipulations_1 = __importDefault(require("./cancelActiveManipulations"));
exports.cancelActiveManipulations = cancelActiveManipulations_1.default;
const Synchronizer_1 = __importDefault(require("./SynchronizerManager/Synchronizer"));
exports.Synchronizer = Synchronizer_1.default;
const svgNodeCache_1 = __importDefault(require("./svgNodeCache"));
exports.svgNodeCache = svgNodeCache_1.default;
const state_1 = __importDefault(require("./state"));
exports.state = state_1.default;
const ToolGroupManager = __importStar(require("./ToolGroupManager"));
exports.ToolGroupManager = ToolGroupManager;
const SynchronizerManager = __importStar(require("./SynchronizerManager"));
exports.SynchronizerManager = SynchronizerManager;
