// @cornerstonejs/tools/dist/cjs/stateManagement/segmentation/convertVolumeToStackSegmentation.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateStackSegmentationState = exports.convertVolumeToStackSegmentation = exports.computeStackSegmentationFromVolume = void 0;
const core_1 = require("@cornerstonejs/core");
const enums_1 = require("../../enums");
const addSegmentationRepresentations_1 = __importDefault(require("./addSegmentationRepresentations"));
const segmentation_1 = require("../../utilities/segmentation");
const segmentationState_1 = require("./segmentationState");
const triggerSegmentationEvents_1 = require("./triggerSegmentationEvents");
function computeStackSegmentationFromVolume({ volumeId, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const segmentationVolume = core_1.cache.getVolume(volumeId);
        const hasCachedImages = segmentationVolume.imageCacheOffsetMap.size > 0;
        let isAllImagesCached = false;
        if (hasCachedImages) {
            isAllImagesCached = segmentationVolume.imageIds.every((imageId) => core_1.cache.getImage(imageId));
        }
        const renderingEngine = (0, core_1.getRenderingEngines)()[0];
        const volumeUsedInOtherViewports = renderingEngine
            .getVolumeViewports()
            .find((vp) => vp.hasVolumeId(volumeId));
        segmentationVolume.decache(!volumeUsedInOtherViewports && isAllImagesCached);
        const imageIdReferenceMap = _getImageIdReferenceMapForStackSegmentation(segmentationVolume);
        return { imageIdReferenceMap };
    });
}
exports.computeStackSegmentationFromVolume = computeStackSegmentationFromVolume;
function convertVolumeToStackSegmentation({ segmentationId, options, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const segmentation = (0, segmentationState_1.getSegmentation)(segmentationId);
        const data = segmentation.representationData
            .LABELMAP;
        const { imageIdReferenceMap } = yield computeStackSegmentationFromVolume({
            volumeId: data.volumeId,
        });
        yield updateStackSegmentationState({
            segmentationId,
            toolGroupId: options.toolGroupId,
            imageIdReferenceMap,
            options,
        });
    });
}
exports.convertVolumeToStackSegmentation = convertVolumeToStackSegmentation;
function updateStackSegmentationState({ segmentationId, toolGroupId, imageIdReferenceMap, options, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const segmentation = (0, segmentationState_1.getSegmentation)(segmentationId);
        if (options === null || options === void 0 ? void 0 : options.removeOriginal) {
            const data = segmentation.representationData
                .LABELMAP;
            if (core_1.cache.getVolume(data.volumeId)) {
                core_1.cache.removeVolumeLoadObject(data.volumeId);
            }
            segmentation.representationData.LABELMAP = {
                imageIdReferenceMap,
            };
        }
        else {
            segmentation.representationData.LABELMAP = Object.assign(Object.assign({}, segmentation.representationData.LABELMAP), { imageIdReferenceMap });
        }
        yield (0, addSegmentationRepresentations_1.default)(toolGroupId, [
            {
                segmentationId,
                type: enums_1.SegmentationRepresentations.Labelmap,
            },
        ]);
        (0, segmentation_1.triggerSegmentationRender)(toolGroupId);
        core_1.eventTarget.addEventListenerOnce(enums_1.Events.SEGMENTATION_RENDERED, () => (0, triggerSegmentationEvents_1.triggerSegmentationDataModified)(segmentationId));
    });
}
exports.updateStackSegmentationState = updateStackSegmentationState;
function _getImageIdReferenceMapForStackSegmentation(segmentationVolume) {
    var _a, _b, _c, _d;
    if ((_a = segmentationVolume.additionalDetails) === null || _a === void 0 ? void 0 : _a.imageIdReferenceMap) {
        return segmentationVolume.additionalDetails.imageIdReferenceMap;
    }
    else if (((_b = segmentationVolume.referencedImageIds) === null || _b === void 0 ? void 0 : _b.length) &&
        !segmentationVolume.referencedImageIds[0].startsWith('derived')) {
        const referencedImageIds = segmentationVolume.referencedImageIds;
        const segmentationImageIds = segmentationVolume.imageIds;
        return (0, segmentation_1.createImageIdReferenceMap)(referencedImageIds, [...segmentationImageIds].reverse());
    }
    else {
        const referencedVolumeId = segmentationVolume.referencedVolumeId;
        const referencedVolume = core_1.cache.getVolume(referencedVolumeId);
        if (!referencedVolume) {
            throw new Error('Cannot convert volumetric segmentation without referenced volume to stack segmentation yet');
        }
        if (!((_c = referencedVolume === null || referencedVolume === void 0 ? void 0 : referencedVolume.imageIds) === null || _c === void 0 ? void 0 : _c.length)) {
            throw new Error('Cannot convert volumetric segmentation without imageIds to stack segmentation yet');
        }
        if ((_d = referencedVolume.imageIds) === null || _d === void 0 ? void 0 : _d[0].startsWith('derived')) {
            throw new Error(`Cannot convert volume segmentation that is derived from another segmentation
         to stack segmentation yet, include the additionalDetails.imageIdReferenceMap
         in the volume segmentation in case you need it for the conversion`);
        }
        const referencedImageIds = referencedVolume.imageIds;
        let segmentationImageIdsToUse = segmentationVolume.imageIds;
        if (!(segmentationImageIdsToUse === null || segmentationImageIdsToUse === void 0 ? void 0 : segmentationImageIdsToUse.length)) {
            segmentationImageIdsToUse =
                segmentationVolume.convertToImageSlicesAndCache();
        }
        return (0, segmentation_1.createImageIdReferenceMap)(referencedImageIds, [...segmentationImageIdsToUse].reverse());
    }
}
