// @cornerstonejs/streaming-image-volume-loader/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enums = exports.helpers = exports.StreamingDynamicImageVolume = exports.StreamingImageVolume = exports.cornerstoneStreamingDynamicImageVolumeLoader = exports.cornerstoneStreamingImageVolumeLoader = void 0;
const cornerstoneStreamingImageVolumeLoader_1 = __importDefault(require("./cornerstoneStreamingImageVolumeLoader"));
exports.cornerstoneStreamingImageVolumeLoader = cornerstoneStreamingImageVolumeLoader_1.default;
const cornerstoneStreamingDynamicImageVolumeLoader_1 = __importDefault(require("./cornerstoneStreamingDynamicImageVolumeLoader"));
exports.cornerstoneStreamingDynamicImageVolumeLoader = cornerstoneStreamingDynamicImageVolumeLoader_1.default;
const StreamingImageVolume_1 = __importDefault(require("./StreamingImageVolume"));
exports.StreamingImageVolume = StreamingImageVolume_1.default;
const StreamingDynamicImageVolume_1 = __importDefault(require("./StreamingDynamicImageVolume"));
exports.StreamingDynamicImageVolume = StreamingDynamicImageVolume_1.default;
const getDynamicVolumeInfo_1 = __importDefault(require("./helpers/getDynamicVolumeInfo"));
const Enums = __importStar(require("./enums"));
exports.Enums = Enums;
const helpers = {
    getDynamicVolumeInfo: getDynamicVolumeInfo_1.default,
};
exports.helpers = helpers;
