// @cornerstonejs/tools/dist/cjs/drawingSvg/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setNewAttributesIfValid = exports.setAttributesIfNecessary = exports.drawRedactionRect = exports.drawArrow = exports.drawTextBox = exports.drawRectByCoordinates = exports.drawRect = exports.drawLinkedTextBox = exports.drawPath = exports.drawPolyline = exports.drawHeight = exports.drawLine = exports.drawHandle = exports.drawHandles = exports.drawEllipseByCoordinates = exports.drawEllipse = exports.drawCircle = exports.draw = void 0;
const draw_1 = __importDefault(require("./draw"));
exports.draw = draw_1.default;
const drawCircle_1 = __importDefault(require("./drawCircle"));
exports.drawCircle = drawCircle_1.default;
const drawEllipse_1 = __importDefault(require("./drawEllipse"));
exports.drawEllipse = drawEllipse_1.default;
const drawEllipseByCoordinates_1 = __importDefault(require("./drawEllipseByCoordinates"));
exports.drawEllipseByCoordinates = drawEllipseByCoordinates_1.default;
const drawHandles_1 = __importDefault(require("./drawHandles"));
exports.drawHandles = drawHandles_1.default;
const drawHandle_1 = __importDefault(require("./drawHandle"));
exports.drawHandle = drawHandle_1.default;
const drawLine_1 = __importDefault(require("./drawLine"));
exports.drawLine = drawLine_1.default;
const drawHeight_1 = __importDefault(require("./drawHeight"));
exports.drawHeight = drawHeight_1.default;
const drawPolyline_1 = __importDefault(require("./drawPolyline"));
exports.drawPolyline = drawPolyline_1.default;
const drawPath_1 = __importDefault(require("./drawPath"));
exports.drawPath = drawPath_1.default;
const drawLinkedTextBox_1 = __importDefault(require("./drawLinkedTextBox"));
exports.drawLinkedTextBox = drawLinkedTextBox_1.default;
const drawRect_1 = __importDefault(require("./drawRect"));
exports.drawRect = drawRect_1.default;
const drawRectByCoordinates_1 = __importDefault(require("./drawRectByCoordinates"));
exports.drawRectByCoordinates = drawRectByCoordinates_1.default;
const drawTextBox_1 = __importDefault(require("./drawTextBox"));
exports.drawTextBox = drawTextBox_1.default;
const drawArrow_1 = __importDefault(require("./drawArrow"));
exports.drawArrow = drawArrow_1.default;
const drawRedactionRect_1 = __importDefault(require("./drawRedactionRect"));
exports.drawRedactionRect = drawRedactionRect_1.default;
const setAttributesIfNecessary_1 = __importDefault(require("./setAttributesIfNecessary"));
exports.setAttributesIfNecessary = setAttributesIfNecessary_1.default;
const setNewAttributesIfValid_1 = __importDefault(require("./setNewAttributesIfValid"));
exports.setNewAttributesIfValid = setNewAttributesIfValid_1.default;
