// @cornerstonejs/core/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setUseSharedArrayBuffer = exports.getShouldUseSharedArrayBuffer = exports.resetUseCPURendering = exports.setPreferSizeOverAccuracy = exports.setUseCPURendering = exports.getShouldUseCPURendering = exports.imageLoadPoolManager = exports.imageRetrievalPoolManager = exports.requestPoolManager = exports.addImageSlicesToViewports = exports.addVolumesToViewports = exports.setVolumesForViewports = exports.utilities = exports.metaData = exports.volumeLoader = exports.registerImageLoader = exports.imageLoader = exports.triggerEvent = exports.eventTarget = exports.cache = exports.createVolumeMapper = exports.getOrCreateCanvas = exports.createVolumeActor = exports.getEnabledElementByViewportId = exports.getEnabledElements = exports.getEnabledElementByIds = exports.getEnabledElement = exports.getRenderingEngines = exports.getRenderingEngine = exports.Surface = exports.ImageVolume = exports.RenderingEngine = exports.WSIViewport = exports.VideoViewport = exports.StackViewport = exports.Viewport = exports.VolumeViewport3D = exports.VolumeViewport = exports.BaseVolumeViewport = exports.Settings = exports.EVENTS = exports.CONSTANTS = exports.Enums = exports.canRenderFloatTextures = exports.getWebWorkerManager = exports.setConfiguration = exports.getConfiguration = exports.peerImport = exports.isCornerstoneInitialized = exports.init = void 0;
exports.ProgressiveRetrieveImages = exports.geometryLoader = exports.resetUseSharedArrayBuffer = void 0;
const Enums = __importStar(require("./enums"));
exports.Enums = Enums;
const CONSTANTS = __importStar(require("./constants"));
exports.CONSTANTS = CONSTANTS;
const enums_1 = require("./enums");
Object.defineProperty(exports, "EVENTS", { enumerable: true, get: function () { return enums_1.Events; } });
const RenderingEngine_1 = require("./RenderingEngine");
Object.defineProperty(exports, "createVolumeActor", { enumerable: true, get: function () { return RenderingEngine_1.createVolumeActor; } });
Object.defineProperty(exports, "createVolumeMapper", { enumerable: true, get: function () { return RenderingEngine_1.createVolumeMapper; } });
Object.defineProperty(exports, "getOrCreateCanvas", { enumerable: true, get: function () { return RenderingEngine_1.getOrCreateCanvas; } });
const RenderingEngine_2 = __importDefault(require("./RenderingEngine"));
exports.RenderingEngine = RenderingEngine_2.default;
const VolumeViewport_1 = __importDefault(require("./RenderingEngine/VolumeViewport"));
exports.VolumeViewport = VolumeViewport_1.default;
const VolumeViewport3D_1 = __importDefault(require("./RenderingEngine/VolumeViewport3D"));
exports.VolumeViewport3D = VolumeViewport3D_1.default;
const BaseVolumeViewport_1 = __importDefault(require("./RenderingEngine/BaseVolumeViewport"));
exports.BaseVolumeViewport = BaseVolumeViewport_1.default;
const StackViewport_1 = __importDefault(require("./RenderingEngine/StackViewport"));
exports.StackViewport = StackViewport_1.default;
const VideoViewport_1 = __importDefault(require("./RenderingEngine/VideoViewport"));
exports.VideoViewport = VideoViewport_1.default;
const WSIViewport_1 = __importDefault(require("./RenderingEngine/WSIViewport"));
exports.WSIViewport = WSIViewport_1.default;
const Viewport_1 = __importDefault(require("./RenderingEngine/Viewport"));
exports.Viewport = Viewport_1.default;
const eventTarget_1 = __importDefault(require("./eventTarget"));
exports.eventTarget = eventTarget_1.default;
const getRenderingEngine_1 = require("./RenderingEngine/getRenderingEngine");
Object.defineProperty(exports, "getRenderingEngine", { enumerable: true, get: function () { return getRenderingEngine_1.getRenderingEngine; } });
Object.defineProperty(exports, "getRenderingEngines", { enumerable: true, get: function () { return getRenderingEngine_1.getRenderingEngines; } });
const cache_1 = __importStar(require("./cache"));
exports.cache = cache_1.default;
Object.defineProperty(exports, "ImageVolume", { enumerable: true, get: function () { return cache_1.ImageVolume; } });
Object.defineProperty(exports, "Surface", { enumerable: true, get: function () { return cache_1.Surface; } });
const imageRetrievalPoolManager_1 = __importDefault(require("./requestPool/imageRetrievalPoolManager"));
exports.imageRetrievalPoolManager = imageRetrievalPoolManager_1.default;
const imageLoadPoolManager_1 = __importDefault(require("./requestPool/imageLoadPoolManager"));
exports.requestPoolManager = imageLoadPoolManager_1.default;
exports.imageLoadPoolManager = imageLoadPoolManager_1.default;
const getEnabledElement_1 = __importStar(require("./getEnabledElement"));
exports.getEnabledElement = getEnabledElement_1.default;
Object.defineProperty(exports, "getEnabledElementByIds", { enumerable: true, get: function () { return getEnabledElement_1.getEnabledElementByIds; } });
Object.defineProperty(exports, "getEnabledElementByViewportId", { enumerable: true, get: function () { return getEnabledElement_1.getEnabledElementByViewportId; } });
Object.defineProperty(exports, "getEnabledElements", { enumerable: true, get: function () { return getEnabledElement_1.getEnabledElements; } });
const metaData = __importStar(require("./metaData"));
exports.metaData = metaData;
const init_1 = require("./init");
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return init_1.init; } });
Object.defineProperty(exports, "getShouldUseCPURendering", { enumerable: true, get: function () { return init_1.getShouldUseCPURendering; } });
Object.defineProperty(exports, "getShouldUseSharedArrayBuffer", { enumerable: true, get: function () { return init_1.getShouldUseSharedArrayBuffer; } });
Object.defineProperty(exports, "isCornerstoneInitialized", { enumerable: true, get: function () { return init_1.isCornerstoneInitialized; } });
Object.defineProperty(exports, "setUseCPURendering", { enumerable: true, get: function () { return init_1.setUseCPURendering; } });
Object.defineProperty(exports, "setPreferSizeOverAccuracy", { enumerable: true, get: function () { return init_1.setPreferSizeOverAccuracy; } });
Object.defineProperty(exports, "setUseSharedArrayBuffer", { enumerable: true, get: function () { return init_1.setUseSharedArrayBuffer; } });
Object.defineProperty(exports, "resetUseCPURendering", { enumerable: true, get: function () { return init_1.resetUseCPURendering; } });
Object.defineProperty(exports, "resetUseSharedArrayBuffer", { enumerable: true, get: function () { return init_1.resetUseSharedArrayBuffer; } });
Object.defineProperty(exports, "getConfiguration", { enumerable: true, get: function () { return init_1.getConfiguration; } });
Object.defineProperty(exports, "setConfiguration", { enumerable: true, get: function () { return init_1.setConfiguration; } });
Object.defineProperty(exports, "getWebWorkerManager", { enumerable: true, get: function () { return init_1.getWebWorkerManager; } });
Object.defineProperty(exports, "canRenderFloatTextures", { enumerable: true, get: function () { return init_1.canRenderFloatTextures; } });
Object.defineProperty(exports, "peerImport", { enumerable: true, get: function () { return init_1.peerImport; } });
const Settings_1 = __importDefault(require("./Settings"));
exports.Settings = Settings_1.default;
const volumeLoader = __importStar(require("./loaders/volumeLoader"));
exports.volumeLoader = volumeLoader;
const imageLoader = __importStar(require("./loaders/imageLoader"));
exports.imageLoader = imageLoader;
const geometryLoader = __importStar(require("./loaders/geometryLoader"));
exports.geometryLoader = geometryLoader;
const ProgressiveRetrieveImages_1 = __importDefault(require("./loaders/ProgressiveRetrieveImages"));
exports.ProgressiveRetrieveImages = ProgressiveRetrieveImages_1.default;
const utilities = __importStar(require("./utilities"));
exports.utilities = utilities;
const imageLoader_1 = require("./loaders/imageLoader");
Object.defineProperty(exports, "registerImageLoader", { enumerable: true, get: function () { return imageLoader_1.registerImageLoader; } });
const triggerEvent_1 = __importDefault(require("./utilities/triggerEvent"));
exports.triggerEvent = triggerEvent_1.default;
const helpers_1 = require("./RenderingEngine/helpers");
Object.defineProperty(exports, "setVolumesForViewports", { enumerable: true, get: function () { return helpers_1.setVolumesForViewports; } });
Object.defineProperty(exports, "addVolumesToViewports", { enumerable: true, get: function () { return helpers_1.addVolumesToViewports; } });
Object.defineProperty(exports, "addImageSlicesToViewports", { enumerable: true, get: function () { return helpers_1.addImageSlicesToViewports; } });
