// @cornerstonejs/core/dist/esm/utilities/loadImageToCanvas.js
import { loadAndCacheImage } from '../loaders/imageLoader';
import * as metaData from '../metaData';
import { RequestType } from '../enums';
import imageLoadPoolManager from '../requestPool/imageLoadPoolManager';
import renderToCanvasGPU from './renderToCanvasGPU';
import renderToCanvasCPU from './renderToCanvasCPU';
import { getConfiguration } from '../init';
export default function loadImageToCanvas(options) {
    const { canvas, imageId, requestType = RequestType.Thumbnail, priority = -5, renderingEngineId = '_thumbnails', useCPURendering = false, thumbnail = false, imageAspect = false, physicalPixels = false, viewportOptions, } = options;
    const devicePixelRatio = window.devicePixelRatio || 1;
    const renderFn = useCPURendering ? renderToCanvasCPU : renderToCanvasGPU;
    return new Promise((resolve, reject) => {
        function successCallback(image, imageId) {
            const { modality } = metaData.get('generalSeriesModule', imageId) || {};
            image.isPreScaled = image.isPreScaled || image.preScale?.scaled;
            if (thumbnail) {
                canvas.height = 256;
                canvas.width = 256;
            }
            if (physicalPixels) {
                canvas.width = canvas.offsetWidth * devicePixelRatio;
                canvas.height = canvas.offsetHeight * devicePixelRatio;
            }
            if (imageAspect) {
                canvas.width = (canvas.height * image.width) / image.height;
            }
            renderFn(canvas, image, modality, renderingEngineId, viewportOptions).then(() => {
                resolve(imageId);
            });
        }
        function errorCallback(error, imageId) {
            console.error(error, imageId);
            reject(error);
        }
        function sendRequest(imageId, imageIdIndex, options) {
            return loadAndCacheImage(imageId, options).then((image) => {
                successCallback.call(this, image, imageId);
            }, (error) => {
                errorCallback.call(this, error, imageId);
            });
        }
        const { useNorm16Texture, preferSizeOverAccuracy } = getConfiguration().rendering;
        const useNativeDataType = useNorm16Texture || preferSizeOverAccuracy;
        const options = {
            targetBuffer: {
                type: useNativeDataType ? undefined : 'Float32Array',
            },
            preScale: {
                enabled: true,
            },
            useNativeDataType,
            useRGBA: !!useCPURendering,
            requestType,
        };
        imageLoadPoolManager.addRequest(sendRequest.bind(null, imageId, null, options), requestType, { imageId }, priority);
    });
}
