// @cornerstonejs/tools/dist/esm/tools/displayTools/SegmentationDisplayTool.js
import { getEnabledElementByIds, utilities as csUtils, } from '@cornerstonejs/core';
import Representations from '../../enums/SegmentationRepresentations';
import { config as segmentationConfig } from '../../stateManagement/segmentation';
import { setSegmentationVisibility } from '../../stateManagement/segmentation/config/segmentationVisibility';
import { getSegmentationRepresentations } from '../../stateManagement/segmentation/segmentationState';
import { getToolGroup } from '../../store/ToolGroupManager';
import { BaseTool } from '../base';
import { surfaceDisplay } from './Surface';
import { contourDisplay } from './Contour';
import { labelmapDisplay } from './Labelmap';
import SegmentationRepresentations from '../../enums/SegmentationRepresentations';
import { addTool, state } from '../../store';
import PlanarFreehandContourSegmentationTool from '../annotation/PlanarFreehandContourSegmentationTool';
const planarContourToolName = PlanarFreehandContourSegmentationTool.toolName;
class SegmentationDisplayTool extends BaseTool {
    constructor(toolProps = {}, defaultToolProps = {
        configuration: {},
    }) {
        super(toolProps, defaultToolProps);
        this.renderSegmentation = (toolGroupId) => {
            const toolGroup = getToolGroup(toolGroupId);
            if (!toolGroup) {
                return;
            }
            const toolGroupSegmentationRepresentations = getSegmentationRepresentations(toolGroupId);
            if (!toolGroupSegmentationRepresentations ||
                toolGroupSegmentationRepresentations.length === 0) {
                return;
            }
            const toolGroupViewports = toolGroup.viewportsInfo.map(({ renderingEngineId, viewportId }) => {
                const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
                if (enabledElement) {
                    return enabledElement.viewport;
                }
            });
            const segmentationRenderList = toolGroupSegmentationRepresentations.map((representation) => {
                const config = this._getMergedRepresentationsConfig(toolGroupId);
                const viewportsRenderList = [];
                const renderers = {
                    [Representations.Labelmap]: labelmapDisplay,
                    [Representations.Contour]: contourDisplay,
                    [Representations.Surface]: surfaceDisplay,
                };
                if (representation.type === SegmentationRepresentations.Contour) {
                    this.addPlanarFreeHandToolIfAbsent(toolGroupId);
                }
                const display = renderers[representation.type];
                for (const viewport of toolGroupViewports) {
                    const renderedViewport = display.render(viewport, representation, config);
                    viewportsRenderList.push(renderedViewport);
                }
                return viewportsRenderList;
            });
            Promise.allSettled(segmentationRenderList).then(() => {
                toolGroupViewports.forEach((viewport) => {
                    viewport.render();
                });
            });
        };
    }
    onSetToolEnabled() {
        const toolGroupId = this.toolGroupId;
        const toolGroupSegmentationRepresentations = getSegmentationRepresentations(toolGroupId);
        if (!toolGroupSegmentationRepresentations ||
            toolGroupSegmentationRepresentations.length === 0) {
            return;
        }
        toolGroupSegmentationRepresentations.forEach((segmentationRepresentation) => {
            setSegmentationVisibility(toolGroupId, segmentationRepresentation.segmentationRepresentationUID, true);
        });
    }
    onSetToolDisabled() {
        const toolGroupId = this.toolGroupId;
        const toolGroupSegmentationRepresentations = getSegmentationRepresentations(toolGroupId);
        if (!toolGroupSegmentationRepresentations ||
            toolGroupSegmentationRepresentations.length === 0) {
            return;
        }
        toolGroupSegmentationRepresentations.forEach((segmentationRepresentation) => {
            setSegmentationVisibility(toolGroupId, segmentationRepresentation.segmentationRepresentationUID, false);
        });
    }
    addPlanarFreeHandToolIfAbsent(toolGroupId) {
        if (!(planarContourToolName in state.tools)) {
            addTool(PlanarFreehandContourSegmentationTool);
        }
        const toolGroup = getToolGroup(toolGroupId);
        if (!toolGroup.hasTool(planarContourToolName)) {
            toolGroup.addTool(planarContourToolName);
            toolGroup.setToolPassive(planarContourToolName);
        }
    }
    _getMergedRepresentationsConfig(toolGroupId) {
        const toolGroupConfig = segmentationConfig.getToolGroupSpecificConfig(toolGroupId);
        const globalConfig = segmentationConfig.getGlobalConfig();
        const mergedConfig = csUtils.deepMerge(globalConfig, toolGroupConfig);
        return mergedConfig;
    }
}
SegmentationDisplayTool.toolName = 'SegmentationDisplay';
export default SegmentationDisplayTool;
