// @cornerstonejs/tools/dist/cjs/stateManagement/annotation/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnotationGroup = exports.FrameOfReferenceSpecificAnnotationManager = exports.visibility = exports.state = exports.selection = exports.locking = exports.config = void 0;
const config = __importStar(require("./config"));
exports.config = config;
const locking = __importStar(require("./annotationLocking"));
exports.locking = locking;
const selection = __importStar(require("./annotationSelection"));
exports.selection = selection;
const state = __importStar(require("./annotationState"));
exports.state = state;
const visibility = __importStar(require("./annotationVisibility"));
exports.visibility = visibility;
const FrameOfReferenceSpecificAnnotationManager_1 = __importDefault(require("./FrameOfReferenceSpecificAnnotationManager"));
exports.FrameOfReferenceSpecificAnnotationManager = FrameOfReferenceSpecificAnnotationManager_1.default;
const AnnotationGroup_1 = __importDefault(require("./AnnotationGroup"));
exports.AnnotationGroup = AnnotationGroup_1.default;
