// @kitware/vtk.js/Utilities/XMLConverter/chemistry-mapper/elements.json.js
var H = [
	1
];
var He = [
	2
];
var Li = [
	3
];
var Be = [
	4
];
var B = [
	5
];
var C = [
	6
];
var N = [
	7
];
var O = [
	8
];
var F = [
	9
];
var Ne = [
	10
];
var Na = [
	11
];
var Mg = [
	12
];
var Al = [
	13
];
var Si = [
	14
];
var P = [
	15
];
var S = [
	16
];
var Cl = [
	17
];
var Ar = [
	18
];
var K = [
	19
];
var Ca = [
	20
];
var Sc = [
	21
];
var Ti = [
	22
];
var V = [
	23
];
var Cr = [
	24
];
var Mn = [
	25
];
var Fe = [
	26
];
var Co = [
	27
];
var Ni = [
	28
];
var Cu = [
	29
];
var Zn = [
	30
];
var Ga = [
	31
];
var Ge = [
	32
];
var As = [
	33
];
var Se = [
	34
];
var Br = [
	35
];
var Kr = [
	36
];
var Rb = [
	37
];
var Sr = [
	38
];
var Y = [
	39
];
var Zr = [
	40
];
var Nb = [
	41
];
var Mo = [
	42
];
var Tc = [
	43
];
var Ru = [
	44
];
var Rh = [
	45
];
var Pd = [
	46
];
var Ag = [
	47
];
var Cd = [
	48
];
var In = [
	49
];
var Sn = [
	50
];
var Sb = [
	51
];
var Te = [
	52
];
var I = [
	53
];
var Xe = [
	54
];
var Cs = [
	55
];
var Ba = [
	56
];
var La = [
	57
];
var Ce = [
	58
];
var Pr = [
	59
];
var Nd = [
	60
];
var Pm = [
	61
];
var Sm = [
	62
];
var Eu = [
	63
];
var Gd = [
	64
];
var Tb = [
	65
];
var Dy = [
	66
];
var Ho = [
	67
];
var Er = [
	68
];
var Tm = [
	69
];
var Yb = [
	70
];
var Lu = [
	71
];
var Hf = [
	72
];
var Ta = [
	73
];
var W = [
	74
];
var Re = [
	75
];
var Os = [
	76
];
var Ir = [
	77
];
var Pt = [
	78
];
var Au = [
	79
];
var Hg = [
	80
];
var Tl = [
	81
];
var Pb = [
	82
];
var Bi = [
	83
];
var Po = [
	84
];
var At = [
	85
];
var Rn = [
	86
];
var Fr = [
	87
];
var Ra = [
	88
];
var Ac = [
	89
];
var Th = [
	90
];
var Pa = [
	91
];
var U = [
	92
];
var Np = [
	93
];
var Pu = [
	94
];
var Am = [
	95
];
var Cm = [
	96
];
var Bk = [
	97
];
var Cf = [
	98
];
var Es = [
	99
];
var Fm = [
	100
];
var Md = [
	101
];
var No = [
	102
];
var Lr = [
	103
];
var Rf = [
	104
];
var Db = [
	105
];
var Sg = [
	106
];
var Bh = [
	107
];
var Hs = [
	108
];
var Mt = [
	109
];
var Ds = [
	110
];
var Rg = [
	111
];
var Cn = [
	112
];
var Uut = [
	113
];
var Uuq = [
	114
];
var Uup = [
	115
];
var Uuh = [
	116
];
var Uus = [
	117
];
var Uuo = [
	118
];
var ATOMS = {
	H: H,
	He: He,
	Li: Li,
	Be: Be,
	B: B,
	C: C,
	N: N,
	O: O,
	F: F,
	Ne: Ne,
	Na: Na,
	Mg: Mg,
	Al: Al,
	Si: Si,
	P: P,
	S: S,
	Cl: Cl,
	Ar: Ar,
	K: K,
	Ca: Ca,
	Sc: Sc,
	Ti: Ti,
	V: V,
	Cr: Cr,
	Mn: Mn,
	Fe: Fe,
	Co: Co,
	Ni: Ni,
	Cu: Cu,
	Zn: Zn,
	Ga: Ga,
	Ge: Ge,
	As: As,
	Se: Se,
	Br: Br,
	Kr: Kr,
	Rb: Rb,
	Sr: Sr,
	Y: Y,
	Zr: Zr,
	Nb: Nb,
	Mo: Mo,
	Tc: Tc,
	Ru: Ru,
	Rh: Rh,
	Pd: Pd,
	Ag: Ag,
	Cd: Cd,
	In: In,
	Sn: Sn,
	Sb: Sb,
	Te: Te,
	I: I,
	Xe: Xe,
	Cs: Cs,
	Ba: Ba,
	La: La,
	Ce: Ce,
	Pr: Pr,
	Nd: Nd,
	Pm: Pm,
	Sm: Sm,
	Eu: Eu,
	Gd: Gd,
	Tb: Tb,
	Dy: Dy,
	Ho: Ho,
	Er: Er,
	Tm: Tm,
	Yb: Yb,
	Lu: Lu,
	Hf: Hf,
	Ta: Ta,
	W: W,
	Re: Re,
	Os: Os,
	Ir: Ir,
	Pt: Pt,
	Au: Au,
	Hg: Hg,
	Tl: Tl,
	Pb: Pb,
	Bi: Bi,
	Po: Po,
	At: At,
	Rn: Rn,
	Fr: Fr,
	Ra: Ra,
	Ac: Ac,
	Th: Th,
	Pa: Pa,
	U: U,
	Np: Np,
	Pu: Pu,
	Am: Am,
	Cm: Cm,
	Bk: Bk,
	Cf: Cf,
	Es: Es,
	Fm: Fm,
	Md: Md,
	No: No,
	Lr: Lr,
	Rf: Rf,
	Db: Db,
	Sg: Sg,
	Bh: Bh,
	Hs: Hs,
	Mt: Mt,
	Ds: Ds,
	Rg: Rg,
	Cn: Cn,
	Uut: Uut,
	Uuq: Uuq,
	Uup: Uup,
	Uuh: Uuh,
	Uus: Uus,
	Uuo: Uuo
};

export { ATOMS as A };
