// @cornerstonejs/tools/dist/cjs/stateManagement/segmentation/helpers/normalizeSegmentationInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const enums_1 = require("../../../enums");
function normalizeSegmentationInput(segmentationInput) {
    var _a, _b;
    const { segmentationId, representation } = segmentationInput;
    const isContourRepresentation = representation.type === enums_1.SegmentationRepresentations.Contour;
    let data = representation.data ? Object.assign({}, representation.data) : null;
    data = !data && isContourRepresentation ? {} : data;
    if (!data) {
        throw new Error('Segmentation representation data may not be undefined');
    }
    if (isContourRepresentation) {
        const contourData = data;
        contourData.geometryIds = (_a = contourData.geometryIds) !== null && _a !== void 0 ? _a : [];
        contourData.annotationUIDsMap = (_b = contourData.annotationUIDsMap) !== null && _b !== void 0 ? _b : new Map();
    }
    return {
        segmentationId,
        cachedStats: {},
        segmentLabels: {},
        label: null,
        segmentsLocked: new Set(),
        type: representation.type,
        activeSegmentIndex: 1,
        representationData: {
            [representation.type]: Object.assign({}, data),
        },
    };
}
exports.default = normalizeSegmentationInput;
